src/richie/apps/courses/admin.py

```diff
@@ -25,6 +38,101 @@ def title(self, obj):
+                self.admin_site.admin_view(self.snapshot),
+                name="cms_course_snapshot",
+            )
+        ] + url_patterns
+
+    @REQUIRE_POST
+    @transaction.atomic
+    def snapshot(self, request, course_id, *args, **kwargs):
+        """
+        Snapshotting a course is making a copy of the course page with all its permissions and
+        extensions, and placing the copy as the first child of the page being snapshotted, then
+        moving all the course run of the page being snapshotted as children of the new snapshot
+        so we keep record of the course as it was when these course runs were played.
+        """
+        try:
+            page = Page.objects.select_related("node__site").get(
+                publisher_is_draft=True, course__id=course_id
+            )
+        except Page.DoesNotExist:
+            return jsonify_request(
+                HttpResponseBadRequest(
+                    force_text(_("Error! Course could not be found."))
+                )
+            )
+
+        # If the page has a parent that is a course page, it is a snaptshot and should therefore
+        # not be allowed to be itself snapshotted.
+        if page.parent_page:
+            try:
+                page.parent_page.course
+            except Course.DoesNotExist:
+                pass
+            else:
+                return jsonify_request(
+                    HttpResponseForbidden(
+                        force_text(_("Error! You can't snapshot a snapshot."))
+                    )
+                )
+
+        site = page.node.site
+        user = request.user
+
+        # User can only snapshot pages he can see
+        can_snapshot_page = page_permissions.user_can_change_page(user, page, site)
+
+        if can_snapshot_page:
+            # User can only snapshot a page if he has the permission to add a page under it.
+            can_snapshot_page = page_permissions.user_can_add_subpage(user, page, site)
+
+        if not can_snapshot_page:
+            return jsonify_request(
+                HttpResponseForbidden(
+                    force_text(
+                        _("Error! You don't have permissions to snapshot this page.")
+                    )
+                )
+            )
+
+        # Copy the page as its own child with its extension.
+        # Titles are set to a timestamp in each language of the original page
+        new_page = page.copy(
+            site, parent_node=page.node, translations=False, extensions=True
+        )
+
+        # The snapshot title and slug is set to a timestamp of the time of snapshot. It is
+        # published only in languages for which the original course page was published.
+        for language in page.get_languages():
+            base = page.get_path(language)
+            timestamp = str(int(time.time()))
+            create_title(
+                language=language,
+                menu_title=timestamp,
+                title=timestamp,
```

Comment:
As we talked about, maybe something a smidge more helpful like `{timestamp} - snapshot of {title.substr(20)...}` or something like that